extractdilkey.c

```diff
@@ -0,0 +1,386 @@
+		{
+			sPubOut = 1;
+		}
+		else if (strcmp(argv[sIdx], "-outraw") == 0)
+		{
+			sRawOut = 1;
+		}
+		else if (strcmp(argv[sIdx], "-inraw") == 0)
+		{
+			sRawIn = 1;
+		}
+        else if (strcmp(argv[sIdx], "-v") == 0)
+        {
+            sVerbose = 1;
+        }
+		else
+		{
+			printf("**** ERROR : Unknown parameter : %s\n", argv[sIdx]);
+			sPrintHelp = 1;
+		}
+	}
+
+
+	if (NULL == sInFile ||
+        (NULL == sOutFile && sRawOut)
+        )
+	{
+		printf("**** ERROR : Invalid input parms\n");
+		sPrintHelp = 1;
+	}
+
+	if (0 != sPrintHelp)
+	{
+		printf("\nextractdilkey -k <input key> [-pubin] [-inraw] [-o <output filename> [-outraw]]\n");
+		exit(0);
+	}
+
+	unsigned char* sRawKey = malloc(BUF_SIZE);
+	unsigned char* sKey = malloc(BUF_SIZE);
+
+	sRc = readFile(sKey,&sKeyBytes,sInFile);
+	if (0 != sRc)
+	{
+		printf("**** ERROR : Unable to read from : %s\n", sInFile);
+		sRc = 1;
+	}
+
+    do {
+    // Now validate our input
+    if (0 == sRc && sPubIn)
+    {
+        if (sRawIn)
+        {
+            // We have a raw Dilithium R2 8x7 public key
+            if (2336 == sKeyBytes)
+            {
+                // We have a raw public key, lets convert it
+                if (sVerbose) printf("extractdilkey: Found raw public key\n");
+                memcpy(sRawKey, sKey, sKeyBytes);
+                sRawBytes = sKeyBytes;
+
+                sKeyBytes = BUF_SIZE;
+
+                // Convert public key
+                sRc = mlca_key2wire(sKey,    sKeyBytes,
+                                    sRawKey, sRawBytes, 0,
+                                    NULL, 0,
+                                    NULL, 0);
+                if (sRc < 0)
+                {
+                    printf("**** ERROR: Failure during public key conversion : %d\n", sRc);
+                    break;
+                }
+                sKeyBytes = sRc;
+                sRc = 0;
+            }
+            else
+            {
+                printf("**** ERROR: Unrecognized raw public key : %s\n", sInFile);
+                sRc = 1;
+                break;
+            }
+        }
+        else
+        {
+            // Attempt to convert encoded key
+            unsigned int sWireType = 0;
+            sRc = mlca_wire2key(sRawKey, sRawBytes, &sWireType,
+                                sKey, sKeyBytes,
+                                NULL, ~0);
+            if (sVerbose) printf("extractdilkey: Found public key\n");
+            // We have a raw Dilithium R2 8x7 public key
+            if (2336 != sRc)
+            {
+                printf("**** ERROR: Unable to convert public key : %d\n",sRc);
+                sRc = 1;
+                break;
+            }
+            else
+            {
+                sRawBytes = sRc;
+                sRc = 0;
+            }
+        }
+
+        if (NULL != sOutFile)
+        {
+            printf("Writing public key to : %s\n", sOutFile);
+            if (sRawOut)
```

Comment:
binary?

Reply:
Raw is not the same as binary.  They are both binary formats